botocore/utils.py

```diff
@@ -636,10 +664,17 @@ def fix_s3_host(request, signature_version, region_name, **kwargs):
     parts = urlsplit(request.url)
     request.auth_path = parts.path
     path_parts = parts.path.split('/')
-    if signature_version in ['s3v4', 'v4']:
-        return
+
+    # Retrieve what the endpoint we will be prepending the bucket name to.
+    if default_endpoint_url is None:
+        default_endpoint_url = parts.netloc
+
     if len(path_parts) > 1:
         bucket_name = path_parts[1]
+        if not bucket_name:
+            # If the bucket name is empty we should not be checking for
+            # dns compatibility.
+            return
         logger.debug('Checking for DNS compatible bucket for: %s',
                      request.url)
         if check_dns_name(bucket_name) and _allowed_region(region_name):
```

Comment:
This is an issue because we're trying to share code between the existing "auto" behavior and "virtual" right?  Because to be clear, I'm talking specifically about the new behavior for "virtual".  So if i have addressing=virtual and a --region of us-gov-west-1, I would expect a DNS compatible bucket to go to:  `bucket.s3-us-gov-west-1.amazonaws.com`.  If the addressing mode is "auto" I'd expect no change in the current behavior.